### PR TITLE
Install vets-website as a submodule into "vagov-apps" directory

### PR DESCRIPTION
**Pull request by ncksllvn**
# Interim CMS
The Interim CMS requires a content-change merged to `master` to be immediately deployed to the production website, via a "CMS" deployment. App-code updates will have to wait for the usual daily, which is referred to as a "full-code" deployment.

## Where we are now
The build for `vets-website` has been adjusted over the past couple sprints to modularize Webpack's build from Metalsmith. This work was prioritized to enable derived-`rootUrl`s for React applications that are moving to a new URL as part of the Brand Consolidation effort. However, we followed an approach that also puts us at a great position for an interim CMS with just some key decisions to make before proceeding.  

I don't anticipate there being many more changes required to our `build.js` either way, but before we do any additional work we should start planning what flags we'll use to indicate full vs. CMS builds, where to find the `file-manifest.json` of precompiled app-code, etc. 

However, before proceeding, the immediate decision to make is...

## Should we split the repo?
**The big decision at this point is whether we should split the `vets-website` repo into `vagov-apps` and `vagov-content`, or leave both content and app-code in `vets-website`.** On the bright side, we're at a great point to make this decision, and neither is conceptually that different, but I think there are ups and downs to each approach.

### Approach 1: Leave vets-website alone
We could leave content and app-code in the same repo, because even though the directories don't reflect the modularized design of app-code and content, they could live under the same roof and coexist just fine.

#### Pros
- No major change to the workflow of content and engineers
- No repo-migration to worry about
- Most likely easier to implement on a short timeline

#### Cons
- Could be difficult to compute the diff of a pull request to determine whether a CMS deploy should be done, or if we should wait for the full-code deploy.
- Changes that require a full-code deploy occurring between CMS deployments could have unknown effects
- Preserving the website state during previous CMS deployments becomes more difficult, as we would have to manually look up what app-code we were linked to at that time (while tracing the lifespan of a production issue, etc), and may be reliant on S3 logs or comparing app-code release dates to find that information.
- Functionally-different pull requests for content and engineers continue to be issued under one roof

### Approach 2: Split vets-website into vagov-content and vagov-apps
As is done with this pull request, we could very easily split `vets-website` like so -

- `vagov-content`
    - `script/`
    - `content`
    - `va-gov/`
    - `assets/`
    - Docker, Jenkins, and other build/config data.

- ` vagov-apps`
    - `src/`
    - `config/`
    - `docs/`
    - App-code configs

**This pull request demonstrates how we would wire the two together.** I wanted to delete the unneeded directories to reflect `vagov-content` above, but that would have made the diff crazy.

#### Pros
- Separation of directory/file-ownership between content and app-code
- No file-diff necessary to determine whether a deployment should be full-code or CMS
    - Changes to `vagov-content` are CMS deployments
    - Changes to `vagov-apps` have to wait for the full-code deployment 
- Could use a submodule to easily preserve app-code version in the content repo

#### Cons
- Repo-migration would require coordination with content and engineers
    - `vets-website` could be left alone, but it would become a submodule (or something conceptually similar) of `vagov-content`, so it would still involve a change in workflow for engineers
    - Content team members would have be instructed to prepare for the repo switch, and any pending pull request or branches in `vets-website` should take that into consideration.
- E2E tests would need to be adapted during automated tests
    - They should belong to `vagov-apps`, but they require the full website, so they would need to run in the context of the parent repo, `vagov-content`. This makes E2E tests complicated.

### Mutual cons
I think both options carry a level of risk with deploying directly to Production during the CMS deployment. I don't think either carries more or less of this risk.

## My opinion - split the repos
I feel inclined to split the repos, because after researching submodules, this seems like a natural use case and also seems to reflect the relationship of how an HTML page (content) imports the JS/CSS assets (app-code). I'm not an expert at this at all, and I don't know Jenkins' limitations, but in my head, our release cycle would look like -

### CMS Deployment
1. `vagov-content` source control is checked out in Jenkins
2. `yarn install` is executed in `vagov-content`
3. `script/build.js` is executed with a `--precompiled {git-sha}` argument
    - This Git SHA is a commit in the `vagov-content` repo of the last full-code deployment. 
4. Content-compilation begins
5. `script/build.js` uses the Git SHA as a flag to skip the app-code compile step
    - The build does not even attempt to import `vagov-apps`, or install any of the `vagov-apps` dependencies, which can be in `vagov-app`'s own `package.json`.
6. `script/build.js` uses the Git SHA to send an HTTP request to find the `file-manifest.json` of precompiled app-code
    - This can exist at `s3://vagov-apps/{{ development | production }}/{{ SHA }}/`
7. Push to S3 bucket

Build finished!

### Full-code Deployment
1. `vagov-content` source control is checked out in Jenkins, with the`--recurse-submodule` flag to indicate that submodules should also be installed.
2. From `vagov-content` -
    1. `git submodule update --remote`
        - This finds the latest Git SHA of the `master` branch in the `vagov-apps` repo, and applies that to the `vagov-apps` submodule.
    2. `git add vagov-apps`
        - Stage the submodule update
    3. `git commit -m "Deploy vagov-apps"`
          - Maybe it would better to do this as a pull request? [Here](https://github.com/department-of-veterans-affairs/vagov-content/pull/4) is an example of what that would look like.
    4. `git push origin master`
3. `yarn install` is executed in `vagov-content`
4. `yarn install` is executed in `vagov-content/vagov-apps`
5. `script/build.js` is executed without the Git SHA
6. Content-compilation begins
7. `script/build.js` compiles the app-code, optimized or development based on build-type.
8. `script/build.js` uses the local `file-manifest.json` to hook up to compiled app-code.
9. Push to S3 bucket

Build Finished!

## Issue tracking
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/13803
https://github.com/department-of-veterans-affairs/vets.gov-team/issues/13013

### CODE REVIEW

File: script/build.js [hunk 1]
```diff
@@ -14,7 +14,7 @@ const permalinks = require('metalsmith-permalinks');
 const sitemap = require('metalsmith-sitemap');
 const watch = require('metalsmith-watch');
 
-const webpackMetalsmithConnect = require('../config/webpack-metalsmith-connect');
+const webpackMetalsmithConnect = require('../vagov-apps/config/webpack-metalsmith-connect');
```
Comment (ncksllvn) [inline script/build.js]: By "modularizing Webpack from Metalsmith", what we mean is that the entire app-code/Webpack compilation is abstracted away from Metalsmith into this single import. The entire Webpack compilation and app-code can easily exist in its own repo, as shown here. `vagov-content` is the top-level repo, with a clone of `vets-website` installed into a directory `vagov-apps` as a submodule. We can now delete the `src` and `config` directories from `vagov-content`. Inversely, `vagov-apps` can safely remove the content directories, build script, etc. This is an example of how splitting the `vets-website` repo could work.

Comment (ncksllvn) [reply script/build.js]: We could also do this inversely, by setting up a repo for `vagov-apps` and converting `vets-website` into `vagov-content`.

**Comment by ncksllvn** [comment]
> I think both options carry a level of risk with deploying directly to Production during the CMS deployment. I don't think either carries more or less of this risk.

I was thinking about this more, and I think there's more risk to not splitting the applications into separate repos. Reason being, during a CMS deployment, the code in `vagov-apps` will not be installed or accessible, so essentially everything there is guarded for full-code deploys only. However, if we leave them together, there seems to be possibility of our build cycle accessing files it shouldn't (by reaching into the `vagov-apps` code) or if we make a miscalculation with our diff and determine a PR to be safe for CMS deployment but that should have waited.

**Comment by jbalboni** [comment]
I talked some more with James about this, I think my opinion is this:

Splitting the repos by moving the va-gov content directory into a separate repo fractures our code in a way that is to cause us problems down the road. Any non-trivial layout, header, footer, or global navigation change typically means changing the JS, Sass, and HTML associated with those features. If we split the repo this way, then the HTML lives in one repo and the JS and Sass lives in another, forcing us to coordinate changes across the two repos. Even if we could find a workflow that makes this not too painful, we've still create a confusing separation between related code.

I think there are two conceptually workable ways of splitting the repo:

1. Pulling out just the va-gov/pages directory (and possibly any other files that are just markdown content)
2. Making vagov-content a fully-fledged site of its own, that controls its own styles and behavior, and make vets-website responsible for just application code

The first option I think will run into some trouble with being able to preview content changes, but I think we could figure out how to make that work. The second option is a ton of work. The first option, for what it's worth, is also more in line with how we proposed a CMS structure for the site could work.

**Comment by jbalboni** [comment]
Nick and I talked through some of this and came up with an adjustment to his current approach that we both think is reasonable.

I wanted to document the workflow and pros/cons of option 1 in my comment above, because it has some advantages compared to the other approach (and some disadvantages). I also prefer it conceptually, though that matters less than what approach is most feasible right now.

In this approach, we would split the files in `va-gov/pages` into a separate repo, likely along with the header and footer json files (which contain links that content folks can update). This repo would be fully owned by content editors, who could make changes and approve them without an ok from engineers. We could have some automated checking in place to make sure files were formatted correctly.

There are a couple approaches to how vets-website would reference the content:

1. A submodule reference
2. A plain text file with the SHA of the associated vagov-content commit, which the build script would use to clone content into a vets-website directory ignored by git
3. Use the github API to fetch the repo at the given SHA (not much different than 2, but easier to adopt to a headless CMS instance later)

When a commit was made in master in the content repo, a webhook would trigger a content deploy on vets-website, which would follow the CMS deployment steps listed by Nick in the description of this PR, except it would happen in the vets-website repo and fetch the content sha/submodule of vagov-content.

Previewing content might be a little tougher to figure out, but it is the same build process as above, except the SHA of the current vagov-content brach/commit would override what was currently checked into master and maybe published to Heroku?

I think the pros of this approach are:

1. Content editors can own a repo that is completely separate from application code
2. There's very little change to the application developer workflow
3. It fits into a headless CMS workflow that we might need to adopt later

Cons:

1. It's harder to enforce that content changes in vagov-content don't break things in vets-website, because it's completely separate. We would likely need to fail builds based on a build process in another repo.
2. Previewing changes is probably harder, because we need to invoke a build process in another repo
3. Just generally, I don't know how feasible this is within Jenkins

@ncksllvn Did I miss any disadvantages of this approach? I feel like there were others.

**Comment by ncksllvn** [comment]
@jbalboni I think the last con would be that we lose that preservation of how the app-code stood at the time of a CMS deployment, because the pages being a child/subdirectory/submodule of the app-code would no longer be able to maintain a pointer to  a commit of the app-code repository in Git. This wouldn't work with a submodule, because there isn't a way to configure a submodule to always take latest. I just tried adding the submodule to the Gitignore 😞 

But I really like the concept of replicating a headless CMS workflow. I think we can actually really simplify things, because if we follow that concept, I don't think we care about that Git SHA described above, because we would instead be embracing the idea of  interfacing with an API or some source of dynamic data. 

To make this work, I think we could imitate the relationship of the `vets-api`  repo with the `vets-api-mockdata` . We could define the `va-gov/pages` directory as the `vagov-content` repo, and clone it into a sibling directory of the `vets-website` repo (later to be renamed `vagov-apps`). This would be the extent of changes for FE engineers to migrate. A CMS deployment would be similar to the above, but slightly different because we no longer need to concern ourselves with updating the submodule, and because `vagov-content` would always require `vets-website` to execute. It would look like -

## CMS Deployment
1. Clone the latest release of `vets-website` into `applications/vets-website`
2. Clone `vagov-content` into `applications/vagov-content`
3. Execute `vet-website/script/build.js` with a flag indicating the location of previously-compiled app-code somewhere in S3
    - `node vagov-apps/script/build.js --cms-deployment {GIT_SHA}`
    - This build would link up to the sibling `vagov-content` directory as the source of Metalsmith pages. Other Metalsmith assets, such as layouts, includes, components, etc. continue to live in `vets-website`.
    - This flag would likely be the commit SHA of the latest release on `vets-website`
4. Skip JS-compilation
5. Grab distant `file-manifest.json` and process it into link references throughout generated HTML
5. Publish to S3

## App/Full-code Deployment
I think it would make sense for `build.js` to remain the single entry point for building the site, because even if we have a separate build script for compiling Webpack, we need to update the hash of each Webpack asset in the script tags of the compiled content pages. It just seems simpler for local work to keep `build.js` as that single point, but either way, separating the Webpack-compilation from the Metalsmith-compilation doesn't have a huge impact.

1. Clone the latest release of `vets-website` into `applications/vets-website`
2. Clone `vagov-content` into `applications/vagov-content`
3. Execute `vet-website/script/build.js` with a flag indicating the destination of the to-be-compiled Webpack app-code
    - `node vagov-apps/script/build.js --webpack-destination {GIT_SHA}`
5. Grab local `file-manifest.json` and process it into link references throughout generated HTML
5. Publish to S3

## Opinion
I think this is my favorite approach, as long as we're willing to let go of the closely-coupled relationship  between the content and the app-code that we're used to on `vets-website`, because we'll no longer be able to easily checkout the prior versions of the website.

**Comment by jbalboni** [comment]
Oh, right, I think this is the other downside, and the reason why you're saying we should skip keeping the vagov-content SHA:

We can't keep a current reference to the git SHA unless we create a new branching structure for vets-website. Because if we're updating vets-website each time we do a content deploy, then we're going to pick up new code committed to master and publish it outside of the normal daily deploy process.

We could get around this with different branching strategies, possibly, but it wouldn't work with our current master branch as production approach.

**Comment by ncksllvn** [comment]
Closing, as we have developed a strategy!

**Pull request closed without merging**